src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/CodeGeneration/CodeGenerationOptions.cs

```diff
@@ -122,7 +126,20 @@ public static CleanCodeGenerationOptions GetCleanCodeGenerationOptions(this IOpt
     public static async ValueTask<CodeGenerationOptions> GetCodeGenerationOptionsAsync(this Document document, CodeGenerationOptions? fallbackOptions, CancellationToken cancellationToken)
     {
         var configOptions = await document.GetAnalyzerConfigOptionsAsync(cancellationToken).ConfigureAwait(false);
-        return configOptions.GetCodeGenerationOptions(document.Project.Services, fallbackOptions);
+
+        if (s_configOptionsToGenerationOptions.TryGetValue(configOptions, out var generationOptions))
+            return generationOptions;
+
+        // Extract into a separate method to avoid async lambda allocation
+        return GetCodeGenerationOptions(document, fallbackOptions, configOptions);
+
+        static CodeGenerationOptions GetCodeGenerationOptions(
+            Document document, CodeGenerationOptions? fallbackOptions, StructuredAnalyzerConfigOptions configOptions)
+        {
+            return s_configOptionsToGenerationOptions.GetValue(
+                configOptions,
+                configOptions => configOptions.GetCodeGenerationOptions(document.Project.Services, fallbackOptions));
+        }
     }
```

Comment:
@tmat i couldn't figure out the best place to introduce caching.  The issue is that once we call into `configOptions.GetCodeGenerationOptions(document.Project.Services, fallbackOptions);` we're into static helper methods that are working with IOptionsReader objects, and we've lost that this is, for example, a StructuredAnalyzerConfigOptions instance.

I didn't feel comfortable caching at a lower layer since it didn't feel clear if it was ok to cache using those as keys.  Here, however, we know we have a strong typed object which represents a snapshot.  

Alternatively, i could make GetCodeGenerationOptions an instance method on StructuredAnalyzerConfigOptions  that cached tha code gen values within it, deferring to the reader apis to construct it on demand.  but i wasn't sure if that was prefered here or not.

Reply:
An alternate approach to this is in https://github.com/dotnet/roslyn/pull/73726.  but i don't know which oen we want to take here.